### PR TITLE
CASSANDRA-16296 Node can join when RF > N in multi-DC

### PR DESCRIPTION
**Pull request by bereng**

### CODE REVIEW

File: bootstrap_test.py [hunk 1]
```diff
@@ -280,6 +280,28 @@ def test_consistent_range_movement_true_with_rf1_should_fail(self):
...
+        node5 = Node('node5',
+                     cluster,
+                     True,
+                     ('127.0.0.5', 9160),
+                     ('127.0.0.5', 7000),
+                     '7500',
+                     '0',
+                     None,
+                     binary_interface=('127.0.0.5', 9042))
+        cluster.add(node5, is_seed=False, data_center="dc1")
```
Comment (bereng) [inline bootstrap_test.py]: Once c16411 merges we should remove "dc1" here as a test for 16411 #justfyi

File: bootstrap_test.py [hunk 2]
```diff
@@ -280,6 +280,28 @@ def test_consistent_range_movement_true_with_rf1_should_fail(self):
...
+        cluster = self.cluster
+        cluster.set_environment_variable('CASSANDRA_TOKEN_PREGENERATION_DISABLED', 'True')
+        cluster.populate([2, 2])
+        cluster.start()
+
+        node1 = cluster.nodelist()[0]
+        session = self.patient_cql_connection(node1)
+        session.execute("CREATE KEYSPACE k WITH REPLICATION = {'class' : 'NetworkTopologyStrategy', 'dc1' : '3'}")
+
+        node5 = Node('node5',
```
Comment (tlasica) [inline bootstrap_test.py]: nit: using parameter names would make it probably easier to read

File: bootstrap_test.py (hunk 1 above)
Comment (tlasica) [inline bootstrap_test.py]: Probably too much, but
should we add a failed read before / successful read after addition?

File: bootstrap_test.py [hunk 3]
```diff
@@ -280,6 +280,28 @@ def test_consistent_range_movement_true_with_rf1_should_fail(self):
     def test_consistent_range_movement_false_with_rf1_should_succeed(self):
         self._bootstrap_test_with_replica_down(False, rf=1)
 
+    def test_RF_GT_Nodes_should_succeed(self):
```
Comment (tlasica) [inline bootstrap_test.py]: is having multiple DCs essential for this problem?
If so then maybe it should be reflected in the test name?

**Review by tlasica — APPROVED** [review]
with some minor questions

File: bootstrap_test.py (hunk 3 above)
Comment (bereng) [reply bootstrap_test.py]: Yep multi-dc is the key. I can change the name.

File: bootstrap_test.py (hunk 1 above)
Comment (bereng) [reply bootstrap_test.py]: It won't hurt...

File: bootstrap_test.py [hunk 4]
```diff
@@ -280,6 +280,33 @@ def test_consistent_range_movement_true_with_rf1_should_fail(self):
...
+        cluster = self.cluster
+        cluster.set_environment_variable('CASSANDRA_TOKEN_PREGENERATION_DISABLED', 'True')
+        cluster.populate([1, 1])
+        cluster.start()
+
+        node1 = cluster.nodelist()[0]
+        session = self.patient_cql_connection(node1)
+        session.execute("CREATE KEYSPACE k WITH REPLICATION = {'class' : 'NetworkTopologyStrategy', 'dc1' : '2'}")
+        session.execute("CREATE TABLE k.testgtrfmultidc (KEY text PRIMARY KEY)")
+        session.execute("INSERT INTO k.testgtrfmultidc (KEY) VALUES ('test_rf_gt_nodes_multidc_should_succeed')")
```
Comment (adelapena) [inline bootstrap_test.py:305]: We could easily verify here that the log warnings are emitted in both nodes, since we don't check that in unit tests:
```python
warning = 'Your replication factor 2 for keyspace k is higher than the number of nodes 1 for datacenter dc1'
if cluster.version() >= '4.0':
    for node in cluster.nodelist():
        assert len(node.grep_log(warning)) == 3
```
Then we could verify that the warning is not emitted for the third node, since we have reached a cluster size that matches the RF:
```python
if cluster.version() >= '4.0':
    assert len(node3.grep_log(warning)) == 0
```

File: bootstrap_test.py [hunk 5]
```diff
@@ -280,6 +280,33 @@ def test_consistent_range_movement_true_with_rf1_should_fail(self):
     def test_consistent_range_movement_false_with_rf1_should_succeed(self):
         self._bootstrap_test_with_replica_down(False, rf=1)
 
+    def test_rf_gt_nodes_multidc_should_succeed(self):
```
Comment (adelapena) [inline bootstrap_test.py:283]: We could add some brief docstring mentioning the purpose of the test and mentioning the related ticket(s) with `@jira_ticket`.

File: bootstrap_test.py (hunk 4 above)
Comment (bereng) [reply bootstrap_test.py:305]: +1

File: bootstrap_test.py [hunk 6]
```diff
@@ -280,6 +280,44 @@ def test_consistent_range_movement_true_with_rf1_should_fail(self):
     def test_consistent_range_movement_false_with_rf1_should_succeed(self):
         self._bootstrap_test_with_replica_down(False, rf=1)
 
+    """
+    Test creating a KS with RF > N on multi DC doesn't break bootstrap
+    @jira_ticket CASSANDRA-16296 CASSANDRA16411
+    """
+    def test_rf_gt_nodes_multidc_should_succeed(self):
```
Comment (adelapena) [inline bootstrap_test.py]: Nit: I think doctring should be the first statement in the function's body. There is also a missed hyphen in the second ticket id:
```suggestion
    def test_rf_gt_nodes_multidc_should_succeed(self):
        """
        Test creating a KS with RF > N on multi DC doesn't break bootstrap
        @jira_ticket CASSANDRA-16296 CASSANDRA-16411
        """
```

File: bootstrap_test.py [hunk 7]
```diff
@@ -280,6 +280,53 @@ def test_consistent_range_movement_true_with_rf1_should_fail(self):
...
+        cluster = self.cluster
+        cluster.set_environment_variable('CASSANDRA_TOKEN_PREGENERATION_DISABLED', 'True')
+        cluster.populate([1, 1])
+        cluster.start()
+
+        node1 = cluster.nodelist()[0]
+        session = self.patient_cql_connection(node1)
+        session.execute("CREATE KEYSPACE k WITH REPLICATION = {'class' : 'NetworkTopologyStrategy', 'dc1' : '3'}")
+        warning = 'Your replication factor 3 for keyspace k is higher than the number of nodes 1 for datacenter dc1'
+        warnings_total = 0
```
Comment (adelapena) [inline bootstrap_test.py]: Nit: this could be into the `if` block body. The same applies for the second initialization of `warnings_total` below.

File: bootstrap_test.py [hunk 8]
```diff
@@ -280,6 +280,53 @@ def test_consistent_range_movement_true_with_rf1_should_fail(self):
     def test_consistent_range_movement_false_with_rf1_should_succeed(self):
         self._bootstrap_test_with_replica_down(False, rf=1)
 
+    def test_rf_gt_nodes_multidc_should_succeed(self):
+        """
+        Test creating a KS with RF > N on multi DC doesn't break bootstrap
```
Comment (adelapena) [inline bootstrap_test.py]: Nit: now it's creating and altering a keyspace.

**Review by adelapena — APPROVED** [review]
The new test looks good to me, I have just left a couple of trivial suggestions.

File: bootstrap_test.py (hunk 8 above)
Comment (bereng) [reply bootstrap_test.py]: The gist imo is about KS creation. The `Alter` bit is just operational commands you do to it after creation. It's relevant to the warnings logic but not to the main problem. Would you agree?

File: bootstrap_test.py [hunk 9]
```diff
@@ -280,6 +280,55 @@ def test_consistent_range_movement_true_with_rf1_should_fail(self):
...
+        cluster.populate([1, 1])
+        cluster.start()
+
+        node1 = cluster.nodelist()[0]
+        session = self.patient_cql_connection(node1)
+        session.execute("CREATE KEYSPACE k WITH REPLICATION = {'class' : 'NetworkTopologyStrategy', 'dc1' : '3'}")
+        warning = 'Your replication factor 3 for keyspace k is higher than the number of nodes 1 for datacenter dc1'
+
+        if cluster.version() >= '4.0':
+            warnings_total = 0
```
Comment (tlasica) [inline bootstrap_test.py]:
```
warnings_total = sum([len(node.grep_log(warning) for node in cluster.node_list()])
assert warnings_total == 1
```
but actually what we would like to check is: "warning on a dc1 node, and no warning on dc2 node"
so maybe more explicit:
```
assert len(node1.grep_log(warning)) == 1
assert len(node2.grep_log(warning)) == 0
```

File: bootstrap_test.py [hunk 10]
```diff
@@ -280,6 +280,55 @@ def test_consistent_range_movement_true_with_rf1_should_fail(self):
...
+                warnings_total += len(node.grep_log(warning))
+            assert warnings_total == 1
+
+        session.execute("ALTER KEYSPACE k WITH REPLICATION = {'class' : 'NetworkTopologyStrategy', 'dc1' : '2'}")
+        session.execute("CREATE TABLE k.testgtrfmultidc (KEY text PRIMARY KEY)")
+        session.execute("INSERT INTO k.testgtrfmultidc (KEY) VALUES ('test_rf_gt_nodes_multidc_should_succeed')")
+        warning = 'Your replication factor 2 for keyspace k is higher than the number of nodes 1 for datacenter dc1'
+
+        if cluster.version() >= '4.0':
+            warnings_total = 0
```
Comment (tlasica) [inline bootstrap_test.py]: same here

File: bootstrap_test.py [hunk 11]
```diff
@@ -280,6 +280,55 @@ def test_consistent_range_movement_true_with_rf1_should_fail(self):
...
+                     thrift_interface=('127.0.0.3', 9160),
+                     storage_interface=('127.0.0.3', 7000),
+                     jmx_port='7300',
+                     remote_debug_port='0',
+                     initial_token=None,
+                     binary_interface=('127.0.0.3', 9042))
+        cluster.add(node3, is_seed=False, data_center="dc1")
+        node3.start(wait_for_binary_proto=True);
+        if cluster.version() >= '4.0':
+            assert len(node3.grep_log(warning)) == 0
```
Comment (tlasica) [inline bootstrap_test.py]: this check for warning can be missed because number of nodes changed.
so for example `of nodes 1` will not be found if the warning would be `of nodes 2` (as we added the node).

File: bootstrap_test.py (hunk 8 above)
Comment (tlasica) [reply bootstrap_test.py]: so maybe rather then describing what test does we should describe what it validates?
"Test validating that a KS with RF > N on multi DC doesn not break a bootsrtap"

**Review by tlasica — CHANGES_REQUESTED** [review]

File: bootstrap_test.py (hunk 9 above)
Comment (adelapena) [reply bootstrap_test.py]: Good point, that makes more sense. Just note that the session created with `patient_cql_connection` can use any node as coordinator so doing the asserts that way would be flaky unless we also create the session with `patient_exclusive_cql_connection`. We can also put `warning` inside the `if` block since I think we are not going to reuse it:
```python
node1, node2 = cluster.nodelist()
session = self.patient_exclusive_cql_connection(node1)
session.execute("CREATE KEYSPACE k WITH REPLICATION = {'class' : 'NetworkTopologyStrategy', 'dc1' : '3'}")

if cluster.version() >= '4.0':
    warning = 'Your replication factor 3 for keyspace k is higher than the number of nodes 1 for datacenter dc1'
    assert len(node1.grep_log(warning)) == 1
    assert len(node2.grep_log(warning)) == 0
```

File: bootstrap_test.py (hunk 11 above)
Comment (adelapena) [reply bootstrap_test.py]: We could just mark the logs before adding the node and search in every node for a generic `'Your replication factor'` message, for example something like:
```python
marks = map(lambda n: n.mark_log(), cluster.nodelist())

# add node...

if cluster.version() >= '4.0':
    warning = 'Your replication factor'
    for (node, mark) in zip(cluster.nodelist(), marks):
        assert len(node.grep_log(warning, from_mark=mark)) == 0
```

File: bootstrap_test.py (hunk 8 above)
Comment (adelapena) [reply bootstrap_test.py]: "validating that a KS with RF > N..." sounds good to me.

File: bootstrap_test.py (hunk 9 above)
Comment (bereng) [reply bootstrap_test.py]: Oh I didn't know about the 'exclusive' cql connection gtk.

File: bootstrap_test.py (hunk 11 above)
Comment (bereng) [reply bootstrap_test.py]: Amended but I used a more particular substring for the check to avoid false matches.

File: bootstrap_test.py [hunk 12]
```diff
@@ -280,6 +280,55 @@ def test_consistent_range_movement_true_with_rf1_should_fail(self):
...
+                     remote_debug_port='0',
+                     initial_token=None,
+                     binary_interface=('127.0.0.3', 9042))
+        cluster.add(node3, is_seed=False, data_center="dc1")
+        node3.start(wait_for_binary_proto=True)
+        if cluster.version() >= '4.0':
+            warning = 'is higher than the number of nodes'
+            for (node, mark) in zip(cluster.nodelist(), marks):
+                assert len(node.grep_log(warning, from_mark=mark)) == 0
+
```
Comment (ekaterinadimitrova2) [inline bootstrap_test.py:328]: While testing, I observed that node3 gets flooded with a lot of those(not part of this patch but made me wonder why so many....):
```
WARN  [MigrationStage:1] 2021-02-10 23:32:53,370 MigrationCoordinator.java:429 - Can't send schema pull request: node /127.0.0.1:7000 is down.
WARN  [MigrationStage:1] 2021-02-10 23:32:53,371 MigrationCoordinator.java:429 - Can't send schema pull request: node /127.0.0.2:7000 is down.
```
/cc @adelapena @tlasica

Comment (bereng) [reply bootstrap_test.py:328]: Maybe we could use the no spam logger?

Comment (ekaterinadimitrova2) [reply bootstrap_test.py:328]: I was also wondering about that

**Review by ekaterinadimitrova2 — APPROVED** [review]

**Review by tlasica — APPROVED** [review]

**Comment by michaelsembwever** [comment]
Merged with https://github.com/apache/cassandra-dtest/commit/493ddae492a9a9a47bc484a7dfa75ef86fd3d9b8
